### PR TITLE
Bring mutating optimisations back

### PR DESCRIPTION
**Pull request by DhairyaLGandhi**

### CODE REVIEW

File: src/mutating.jl [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  mt, vt, βt = state
+
+  @. mt = β[1] * mt + (one(T) - β[1]) * dx
+  @. vt = β[2] * vt + (one(T) - β[2]) * dx ^ 2
+  @. dx = mt / (one(T) - βt[1]) / (sqrt(vt / (one(T) - βt[2])) + ϵ) * η
+
+  return dx, (mt, vt, βt .* β)
+end
+
+function apply(o::RADAM, x, dx, state)
```
Comment (darsnack) [inline src/mutating.jl:45]:
```suggestion
function apply!(o::RADAM, x, dx, state)
```

File: src/mutating.jl [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+function apply(o::RMSProp, x, dx, state)
+  η, ρ, ϵ, acc = o.eta, o.rho, o.epsilon, state
+  @. acc = ρ * acc + (1 - ρ) * dx^2
+  @. dx = dx * (η / (sqrt(acc) + ϵ))
+
+  return dx, acc
+end
+
+function apply(o::ADAM{T}, x, dx, state) where T
```
Comment (darsnack) [inline src/mutating.jl:34]:
```suggestion
function apply!(o::ADAM{T}, x, dx, state) where T
```

File: src/mutating.jl [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+function apply(o::Nesterov, x, dx, state)
+  η, ρ, v = o.eta, o.rho, state
+  @. d = ρ^2 * v - (1+ρ) * η * dx
+  @. v = ρ * v - η * dx
+
+  return -d, v
+end
+
+function apply(o::RMSProp, x, dx, state)
```
Comment (darsnack) [inline src/mutating.jl:26]:
```suggestion
function apply!(o::RMSProp, x, dx, state)
```

File: src/mutating.jl [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+end
+
+function apply!(o::Momentum, x, dx, state)
+  η, ρ, v = o.eta, o.rho, state
+  @. v = ρ * v - η * dx
+
+  return -v, v
+end
+
+function apply(o::Nesterov, x, dx, state)
```
Comment (darsnack) [inline src/mutating.jl:18]:
```suggestion
function apply!(o::Nesterov, x, dx, state)
```

**Review by darsnack — CHANGES_REQUESTED** [review]
Some typos

File: src/mutating.jl [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    @. dx = mt / (1 - βt[1]) * η
+  end
+
+  return dx, (mt, vt, βt .* β, t + 1)
+end
+
+function apply!(o::AdaMax, x, dx, state)
+  η, β, ϵ = o.eta, o.beta, o.epsilon
+
+  (mt, ut), βt = state
```
Comment (darsnack) [inline src/mutating.jl:67]:
```suggestion
  mt, ut, βt = state
```

File: src/mutating.jl [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  @. ut = max(β[2] * ut, abs(dx))
+  @. dx = (η/(1 - βt[1])) * mt/(ut + ϵ)
+
+  return dx, ((mt, ut), βt .* β)
+end
+
+function apply!(o::AdaMax, x, dx, state)
+  η, β, ϵ = o.eta, o.beta, o.epsilon
+
+  (mt, ut), βt = state
```
Comment (darsnack) [inline src/mutating.jl:79]:
```suggestion
  mt, ut, βt = state
```

File: src/mutating.jl [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  @. ut = max(β[2] * ut, abs(dx))
+  @. dx = (η/(1 - βt[1])) * mt/(ut + ϵ)
+
+  return dx, ((mt, ut), βt .* β)
+end
+
+function apply!(o::OADAM, x, dx, state)
+  η, β, ϵ = o.eta, o.beta, o.epsilon
+
+  (mt, vt), βt, dx_ = state
```
Comment (darsnack) [inline src/mutating.jl:91]:
```suggestion
  mt, vt, βt, dx_ = state
```

File: src/mutating.jl [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  @. v̂t = max(v̂t, vt)
+  @. dx = η * mt / (sqrt(v̂t) + ϵ)
+
+  return dx, (mt, vt, v̂t)
+end
+
+function apply!(o::NADAM, x, dx, state)
+  η, β, ϵ = o.eta, o.beta, o.epsilon
+
+  (mt, vt), βt = state.decays
```
Comment (darsnack) [inline src/mutating.jl:141]:
```suggestion
  mt, vt, βt = state
```

**Review by darsnack — COMMENTED** [review]
Couple more typos restructuring the state tuple

File: src/mutating.jl [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+function apply!(o::AdaBelief, x, dx, state)
+  η, β, ϵ = o.eta, o.beta, o.epsilon
+  mt, st = state
+
+  @. mt = β[1] * mt + (1 - β[1]) * dx
+  @. st = β[2] * st + (1 - β[2]) * (dx - mt)^2
+  @. dx =  η * mt / (sqrt(st) + ϵ)
+
+  return dx, (mt, st)
+end
```
Comment (darsnack) [inline src/mutating.jl:160]:
```suggestion
end

function apply!(o::OptimiserChain, x, dx, states)
  new_states = similar(states)
  for (i, (opt, state)) in enumerate(zip(o.opts, states))
    _, new_states[i] = apply!(opt, x, dx, state)
  end

  return dx, new_states
end
```

File: src/mutating.jl [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+  η, β, ϵ = o.eta, o.beta, o.epsilon
+
+  mt, vt, βt = state
+
+  @. mt = β[1] * mt + (1 - β[1]) * dx
+  @. vt = β[2] * vt + (1 - β[2]) * dx^2
+  @. dx = (β[1] * mt / (1 - β[1] * βt[1]) + (1 - β[1]) * dx / (1 - βt[1])) / 
+          (sqrt(vt * β[2] / (1 - βt[2])) + ϵ) * η
+
+  return dx, ((mt, vt), βt .* β)
```
Comment (darsnack) [inline src/mutating.jl:148]:
```suggestion
  return dx, (mt, vt, βt .* β)
```

File: src/mutating.jl [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+  mt, vt, βt, dx_ = state
+
+  @. mt = β[1] * mt + (1 - β[1]) * dx
+  @. vt = β[2] * vt + (1 - β[2]) * dx^2
+  @. dx = -dx_
+  @. dx_ = η * mt / (1 - βt[1]) / (sqrt(vt / (1 - βt[2])) + ϵ)
+  @. dx += 2*dx_
+
+  return dx, ((mt, vt), βt .* β, dx_)
```
Comment (darsnack) [inline src/mutating.jl:99]:
```suggestion
  return dx, (mt, vt, βt .* β, dx_)
```

File: src/mutating.jl [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+function apply!(o::AdaMax, x, dx, state)
+  η, β, ϵ = o.eta, o.beta, o.epsilon
+
+  mt, ut, βt = state
+
+  @. mt = β[1] * mt + (1 - β[1]) * dx
+  @. ut = max(β[2] * ut, abs(dx))
+  @. dx = (η/(1 - βt[1])) * mt/(ut + ϵ)
+
+  return dx, ((mt, ut), βt .* β)
```
Comment (darsnack) [inline src/mutating.jl:85]:
```suggestion
  return dx, (mt, ut, βt .* β)
```

File: src/mutating.jl [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
```
Comment (darsnack) [inline src/mutating.jl:73]:
```suggestion
  return dx, (mt, ut, βt .* β)
```

**Comment by darsnack** [comment]
@DhairyaLGandhi I have rebased this locally, do you mind if I push it?

**Comment by DhairyaLGandhi** [comment]
I think I wanted to handle this separately. It also seems like some other changes were merged hastily?

**Comment by darsnack** [comment]
What do you mean handled separately? This has been in need of a rebase for months (any recent changes weren't the reason for preventing this from being merged).

**Comment by DhairyaLGandhi** [comment]
Mutation isn't the goal - so Handling the code duplication separately. Mutation should not be the default unless it is absolutely needed. The api change also needs to be accounted for.

**Comment by darsnack** [comment]
I am confused. Do you still want to merge this PR or not? All I want is to merge this PR.

**Comment by DhairyaLGandhi** [comment]
We can merge it when we need it. We'd prolly not make mutation the default though.

**Comment by darsnack** [comment]
Superseded by #31.

**Pull request closed without merging**